liblepton/scheme/netlist/backend.scm

```diff
@@ -0,0 +1,308 @@
+;;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+(define-module (netlist backend)
+  #:use-module (ice-9 format)
+  #:use-module (ice-9 ftw)
+  #:use-module (srfi srfi-1)
+  #:use-module (srfi srfi-9)
+  #:use-module (srfi srfi-26)
+
+  #:use-module (lepton file-system)
+  #:use-module (lepton gettext)
+  #:use-module (lepton log)
+
+  #:use-module (netlist error)
+  #:use-module (netlist mode)
+
+  #:export (lookup-legacy-backends
+            lookup-module-backends
+            run-backend
+            make-legacy-backend
+            make-module-backend))
+
+(define-record-type <backend>
+  (backend path name runner legacy)
+  backend?
+  (path backend-path set-backend-path!)
+  (name backend-name set-backend-name!)
+  (runner backend-runner set-backend-runner!)
+  (legacy backend-legacy set-backend-legacy!))
+
+
+(define %backend-prefix "gnet-")
+(define %backend-suffix ".scm")
+(define %backend-prefix-length (string-length %backend-prefix))
+(define %backend-suffix-length (string-length %backend-suffix))
+
+
+(define (error-backend-wrong-file-name name)
+  (netlist-error 1 (G_ "Can't load backend file ~S.\n~
+                        Backend files are expected to have names like \"gnet-NAME.scm\"\n~
+                        and contain entry point function NAME (where NAME is the backend's name).\n")
+                 name))
+
+
+(define (error-backend-not-found-by-name backend )
+  (netlist-error 1 (G_ "Could not find backend `~A' in load path.\n~
+                        Run `~A --list-backends' for a full list of available backends.\n")
+                 backend (car (program-arguments))))
+
+(define (error-backend-proc-not-found func-name backend-path)
+  (netlist-error 1 (G_ "Could not find function ~S in backend ~S.\n")
+                 func-name backend-path))
+
+
+(define (legacy-backend-filename? filename)
+  "Return #t if FILENAME is a backend filename, otherwise return
+#f."
+  (and (string-prefix? %backend-prefix filename)
+       (string-suffix? %backend-suffix filename)))
+
+
+(define (module-backend-filename? filename)
+  (and (string-suffix? %backend-suffix filename)
+       (not (legacy-backend-filename? (basename filename)))))
+
+
+(define (backend-filename->proc-name filename)
+  "Transforms FILENAME to a backend name which is also the name of
+the procedure the backend runs.  For legacy backends, the name is
+formed by dropping the prefix \"gnet-\" and the extenstion
+\".scm\".  Returns the resulting string or #f if FILENAME does not
+meet the specified requirements."
+  (and (legacy-backend-filename? filename)
+       (string-drop-right (string-drop filename %backend-prefix-length)
+                          %backend-suffix-length)))
+
+
+(define (search-backend name)
+  "Searches for backend by its NAME."
+  (%search-load-path (format #f "gnet-~A.scm" name)))
+
+
+(define (backend-name-by-path path)
+  (or (backend-filename->proc-name path)
```

Comment:
Should be `( basename path )` here, otherwise file backends (-f) won't work:
`( backend-filename->proc-name (basename path) )`